vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -78,69 +132,176 @@
+                        // this works also with single digit day or month.
+                        // Day or month of less than 10 may still be 2 digit prefixed
+                        // with 0 depending on locale.
+                        if (part === dayPart && date.dayMatch) {
+                            adjust -= 2 - date.dayMatch.length;
+                        } else if (part === monthPart && date.monthMatch) {
+                            adjust -= 2 - date.monthMatch.length;
+                        } else if (part === yearPart) {
+                            return yearPart.index + adjust;
+                        }
+                    }
+                };
+                indexOfYearForFormattedDate = tryCatchWrapper(indexOfYearForFormattedDate);
+
+                /**
+                 * Prefixes the year in the given date string with zeros if
+                 * needed to make it four digits
+                 * @param {string} formattedDate Date formatted with Date.toLocaleDateString()
+                 * @returns {string}
+                 */
+                let yearToFourDigitInFormattedDate = function (formattedDate) {
+                    const date = /** @type {DateHash} */ _parseDate(formattedDate);
+                    if (!date) {
+                        return formattedDate;
+                    }
+                    if (date.yearMatch.length >= 4) {
+                        return formattedDate;
+                    }
+                    const yearIndex = indexOfYearForFormattedDate(date);
+                    const yearStr = String(date.year).replace(/\d+/, y => '0000'.substring(y.length) + y);
+                    const beforeYear = formattedDate.substring(0, yearIndex);
+                    const afterYear = formattedDate.substring(yearIndex + String(date.year).length);
+                    return beforeYear + yearStr + afterYear;
+                }
+                yearToFourDigitInFormattedDate = tryCatchWrapper(yearToFourDigitInFormattedDate);
+
+                /**
+                 * @param {DateHash} date
+                 * @returns {string}
+                 */
+                const formatDate = function (date) {
                     let rawDate = new Date(Date.UTC(date.year, date.month, date.day));
-                    return cleanString(rawDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
-                });
+                    if (date.year >= 0 && date.year <= 99) {
+                        // JS Date constructor converts years 0-99 to 1900-1999 so
+                        // this is needed to fix that when we want an explicit year.
+                        rawDate.setUTCFullYear(date.year);
+                    }
+                    let formattedDate = cleanString(rawDate.toLocaleDateString(locale, { timeZone: 'UTC' }));
+                    if (date.year < 1000) {
+                        formattedDate = yearToFourDigitInFormattedDate(formattedDate);
+                    }
```

Comment:
Fix for years less than 1000 (prefix them with zeros up to 4 digits in `formatDate`).